File: src/python/cram/cramfile.py
##############################################################################
# Copyright (c) 2014, Lawrence Livermore National Security, LLC.
# Produced at the Lawrence Livermore National Laboratory.
#
# This file is part of Cram.
# Written by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
# LLNL-CODE-661100
#
# For details, see https://github.com/scalability-llnl/cram.
# Please also see the LICENSE file for our notice and the LGPL.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License (as published by
# the Free Software Foundation) version 2.1 dated February 1999.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
# conditions of the GNU General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
##############################################################################
"""
This module defines the CramFile class, which is used to store many of
job invocations to be run later in the same MPI job.  A job invocation
is simply the context needed to run a single MPI job: number of
processes, working directory, command-line arguments, and environment.

The format is designed to be read easily from a C program, so it only
writes simple ints and strings.  Ints are all unsigned. Strings start
with an integer length, after which all the characters are written out.

CramFiles use a very simple form of compression to store each job's
environment, since the environment can grow to be very large and is
usually quite redundant.  For each job appended to a CramFile after
the first, we compare its environment to the first job's environment,
and we only store the differences.

We could potentially get more compression out of comparing each
environment to its successor, but that would mean that you'd need to
read all preceding jobs to decode one.  We wanted a format that would
allow scattering jobs very quickly to many MPI processes.

Sample usage:

  cf = CramFile('file.cram', 'w')
  cf.pack(Job(64,             # number of processes
              os.getcwd(),    # working dir, as a string
              sys.argv[1:],   # cmdline args, as a list
              os.env))        # environment, as a dict.
  cf.close()

To read from a CramFile, use len and iterate:

  cf = CramFile('file.cram')
  num_jobs = len(cf)
  for job in cf:
      # do something with job
  cf.close()

Like regular files, CramFiles do not support indexing.  To do that,
read all jobs into a list first, then index the list.

Here is the CramFile format.  '*' below means that the section can be
repeated a variable number of times.

Type       Name
========================================================================
Header
------------------------------------------------------------------------
int(4)       0x6372616d ('cram')
int(4)       Version
int(4)       # of jobs
int(4)       # of processes
int(4)       Size of max job record in this file

* Job records
------------------------------------------------------------------------
  int(4)     Size of job record in bytes
  int(4)     Number of processes
  str        Working dir

  int(4)     Number of command line arguments
   * str       Command line arguments, in original order

  int(4)     Number of subtracted env var names (0 for first record)
   * str       Subtracted env vars in sorted order.
  int(4)     Number of added or changed env vars
   * str      Names of added/changed var
   * str      Corresponding value

Env vars are stored alternating keys and values, in sorted order by key.
========================================================================
"""
import os
import re

from collections import defaultdict
from contextlib import contextmanager, closing

from cram.serialization import *
import llnl.util.tty as tty

# Magic number goes at beginning of file.
_magic = 0x6372616d

# Increment this when the binary format changes (hopefully infrequent)
_version = 2

# Offsets of file header fields
_magic_offset   = 0
_version_offset = 4
_njobs_offset   = 8
_nprocs_offset  = 12
_max_job_offset = 16

# Default name for cram executable.
USE_APP_EXE = "<exe>"


@contextmanager
def save_position(stream):
    """Context for doing something while saving the current position in a
       file."""
    pos = stream.tell()
    yield
    stream.seek(pos)


def compress(base, modified):
    """Compare two dicts, modified and base, and return a diff consisting
       of two parts:

       1. missing: set of keys in base but not in modified
       2. changed: dict of key:value pairs either in modified but not in base,
          or that have different values in base and modified.

       This simple compression scheme is used to compress the environment
       in cramfiles, since that takes up the bulk of the space.
    """
    missing = set(base.keys()).difference(modified)
    changed = { k:v for k,v in modified.items()
                if k not in base or base[k] != v }
    return missing, changed


def decompress(base, missing, changed):
    """Given the base dict and the output of compress(), reconstruct the
       modified dict."""
    d = base.copy()
    for k in missing:
        del d[k]
    d.update(changed)
    return d


class Job(object):
    """Simple class to represent one job invocation packed into a cramfile.
       This contains all environmental context needed to launch the job
       later from within MPI.
    """
    def __init__(self, num_procs, working_dir, args, env):
        """Construct a new Job object.

        Arguments:
        num_procs   -- number of processes to run on.
        working_dir -- path to working directory for job.
        args        -- sequence of arguments, INCLUDING the executable name.
        env         -- dict containng environment.

        """

        self.num_procs = num_procs
        self.working_dir = working_dir

        # Be lenient about the args.  Let the user pass a string if he
        # wants, and split it like the shell would do.
        if isinstance(args, basestring):
            args = re.split(r'\s+', args)
        self.args = args

        self.env = env


    def __eq__(self, other):
        return (self.num_procs == other.num_procs and
                self.working_dir == other.working_dir and
                self.args == other.args and
                self.env == other.env)


    def __ne__(self, other):
        return not (self == other)


class CramFile(object):
    """A CramFile compactly stores a number of Jobs, so that they can
       later be run within the same MPI job by cram.
    """
    def __init__(self, filename, mode='r'):
        """The CramFile constructor functions much like open().

           The constructor takes a filename and an I/O mode, which can
           be 'r', 'w', or 'a', for read, write, or append.

           Opening a CramFile for writing will create a file with a
           simple header containing no jobs.
        """
        # Save the first job from the file.
        self.first_job = None

        self.mode = mode
        if mode not in ('r', 'w', 'a'):
            raise ValueError("Mode must be 'r', 'w', or 'a'.")

        if mode == 'r':
            if not os.path.exists(filename) or os.path.isdir(filename):
                tty.die("No such file: %s" % filename)

            self.stream = open(filename, 'rb')
            self._read_header()

        elif mode == 'w' or (mode == 'a' and not os.path.exists(filename)):
            self.stream = open(filename, 'wb')
            self.version = _version
            self.num_jobs = 0
            self.num_procs = 0
            self.max_job_size = 0
            self._write_header()

        elif mode == 'a':
            self.stream = open(filename, 'rb+')
            self._read_header()
            self.stream.seek(0, os.SEEK_END)


    def _read_header(self):
        """Jump to the beginning of the file and read the header.  The cursor
           will be at the end of the header on completion, so you will need to
           save it if you want to end up somewhere else."""
        self.stream.seek(0)

        magic = read_int(self.stream, 4)
        if magic != _magic:
            raise IOError("%s is not a Cramfile!")

        self.version = read_int(self.stream, 4)
        if self.version != _version:
            raise IOError(
                "Version mismatch: File has version %s, but this is version %s"
                % (self.version, _version))

        self.num_jobs = read_int(self.stream, 4)
        self.num_procs = read_int(self.stream, 4)
        self.max_job_size = read_int(self.stream, 4)

        # read in the first job automatically if it is there, since
        # it is used for compression of subsequent jobs.
        if self.num_jobs > 0:
            self._read_job()


    def _write_header(self):
        """Jump to the beginning of the file and write the header."""
        self.stream.seek(0)
        write_int(self.stream, _magic, 4)
        write_int(self.stream, self.version, 4)
        write_int(self.stream, self.num_jobs, 4)
        write_int(self.stream, self.num_procs, 4)
        write_int(self.stream, self.max_job_size, 4)


    def _pack(self, job):
        """Appends a job to a cram file, compressing the environment in the
           process."""
        if self.mode == 'r':
            raise IOError("Cannot pack into CramFile opened for reading.")

        # Save offset and write 0 for size to start with
        start_offset = self.stream.tell()
        size = 0
        write_int(self.stream, 0, 4)

        # Number of processes
        size += write_int(self.stream, job.num_procs, 4)

        # Working directory
        size += write_string(self.stream, job.working_dir)

        # Command line arguments
        size += write_int(self.stream, len(job.args), 4)
        for arg in job.args:
            size += write_string(self.stream, arg)

        # Compress using first dict
        missing, changed = compress(
            self.first_job.env if self.first_job else {}, job.env)

        # Subtracted env var names
        size += write_int(self.stream, len(missing), 4)
        for key in sorted(missing):
            size += write_string(self.stream, key)

        # Changed environment variables
        size += write_int(self.stream, len(changed), 4)
        for key in sorted(changed.keys()):
            size += write_string(self.stream, key)
            size += write_string(self.stream, changed[key])

        with save_position(self.stream):
            # Update job chunk size
            self.stream.seek(start_offset)
            write_int(self.stream, size, 4)

            # Update total number of jobs in file.
            self.num_jobs += 1
            self.stream.seek(_njobs_offset)
            write_int(self.stream, self.num_jobs, 4)

            # Update total number of processes in all jobs.
            self.num_procs += job.num_procs
            self.stream.seek(_nprocs_offset)
            write_int(self.stream, self.num_procs, 4)

            # Update max job size if necessary
            if size > self.max_job_size:
                self.max_job_size = size
                self.stream.seek(_max_job_offset)
                write_int(self.stream, self.max_job_size, 4)

        # Discard all but hte first job after writing.  This conserves
        # memory when writing cram files.
        if not self.first_job:
            self.first_job = Job(
                job.num_procs, job.working_dir, list(job.args), job.env.copy())


    def pack(self, *args, **kwargs):
        """Pack a Job into a cram file.

        Takes either a Job object, or Job constructor params.
        """
        if len(args) == 1:
            job = args[0]
            if not isinstance(job, Job):
                raise TypeError(
                    "Must pass a Job object, or (nprocs, working_dir, args, env)")
            if kwargs:
                raise TypeError("Cannot pass kwargs with cramfile.pack(Job)")
            self._pack(job)

        elif len(args) == 4:
            nprocs, working_dir, args, env = args

            # Split strings for the caller.
            if isinstance(args, basestring):
                args = re.split(r'\s+', args)

            # By default, cram takes app's exe name.
            exe = kwargs.pop('exe', USE_APP_EXE)
            if kwargs:
                raise ValueError("%s is an invalid keyword arg for this function!"
                                 % next(iter(kwargs.keys())))

            args = [exe] + list(args)
            self._pack(Job(nprocs, working_dir, args, env))


    def _read_job(self):
        """Read the next job out of the CramFile.

           This is an internal method because it's used to load stuff
           that isn't already in memory.  Client code should use
           len(), [], or iterate to read jobs from CramFiles.
        """
        # Size of job record
        job_bytes   = read_int(self.stream, 4)
        start_pos = self.stream.tell()

        # Number of processes
        num_procs   = read_int(self.stream, 4)

        # Working directory
        working_dir = read_string(self.stream)

        # Command line arguments
        num_args    = read_int(self.stream, 4)
        args        = []
        for i in xrange(num_args):
            args.append(read_string(self.stream))

        # Subtracted environment variables
        num_missing = read_int(self.stream, 4)
        missing     = []
        for i in xrange(num_missing):
            missing.append(read_string(self.stream))

        # Changed environment variables
        num_changed = read_int(self.stream, 4)
        changed = {}
        for i in xrange(num_changed):
            key = read_string(self.stream)
            val = read_string(self.stream)
            changed[key] = val

        # validate job record size
        actual_size = self.stream.tell() - start_pos
        if actual_size != job_bytes:
            raise Exception("Cram file job record size is invalid! "+
                            "Expected %d, found %d" % (job_bytes, actual_size))

        # Decompress using first dictionary
        env = decompress(self.first_job.env if self.first_job else {},
                         missing, changed)

        job = Job(num_procs, working_dir, args, env)
        if not self.first_job:
            self.first_job = job
        return job


    def __iter__(self):
        """Iterate over all jobs in the CramFile."""
        if self.mode != 'r':
            raise IOError("Cramfile is not opened for reading.")

        yield self.first_job
        for i in xrange(1, self.num_jobs):
            yield self._read_job()


    def __len__(self):
        """Number of jobs in the file."""
        return self.num_jobs


    def close(self):
        """Close the underlying file stream."""
        self.stream.close()

File: src/c/libcram/cram_file.c
//////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2014, Lawrence Livermore National Security, LLC.
// Produced at the Lawrence Livermore National Laboratory.
//
// This file is part of Cram.
// Written by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
// LLNL-CODE-661100
//
// For details, see https://github.com/scalability-llnl/cram.
// Please also see the LICENSE file for our notice and the LGPL.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License (as published by
// the Free Software Foundation) version 2.1 dated February 1999.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
// conditions of the GNU General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
//////////////////////////////////////////////////////////////////////////////
#include <unistd.h>
#include <stdbool.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <arpa/inet.h>

#include "cram_file.h"

// Magic number goes at beginning of file
#define MAGIC 0x6372616d

// Tag for cram messages
#define CRAM_TAG 7675

// Offsets of file header fields
#define MAGIC_OFFSET       0
#define VERSION_OFFSET     4
#define NJOBS_OFFSET       8
#define NPROCS_OFFSET      12
#define MAX_JOB_OFFSET     16

// offset of first job record
#define JOB_RECORD_OFFSET  20

// max concurrent ranks to send job records to at once.
#define MAX_CONCURRENT_PEERS 512

// Ideal number of bytes to use for Lustre read buffers: 2MB.
#define LUSTRE_BUFFER_SIZE 2097152

// Default cram executable name: means we should use argv[0] for exe.
#define CRAM_DEFAULT_EXE "<exe>"


// ------------------------------------------------------------------------
// Globals used by fortran arg routines
// ------------------------------------------------------------------------
int cram_argc = 0;
const char **cram_argv = NULL;


// ------------------------------------------------------------------------
// Utility functions
// ------------------------------------------------------------------------

///
/// Find strings efficiently in a sorted array using bsearch.
///
static inline size_t index_of(size_t num_elts,
                              const char **sorted_array,
                              const char *string) {
  const char **pos = bsearch(string, sorted_array, num_elts, sizeof(char*),
                             (int(*)(const void*, const void*))strcmp);
  return (pos - sorted_array);
}


///
/// Free all the strings in a char**, then free the array itself.
///
static void free_string_array(int num_elts, const char **arr) {
  for (int i=0; i < num_elts; i++) {
    free((char*)arr[i]);
  }
  free(arr);
}


///
/// Create a copy of an entire string array.
///
static const char **dup_string_array(int num_elts, const char **src) {
  const char **arr = malloc(sizeof(char*) * num_elts);
  for (int i=0; i < num_elts; i++) {
    arr[i] = strdup(src[i]);
  }
  return arr;
}


///
/// Read a cram int from a FILE*.
///
static int file_read_int(const cram_file_t *file) {
  int buf;
  size_t ints = fread(&buf, sizeof(int), 1, file->fd);
  if (ints != 1) {
    fprintf(stderr, "Error reading cram file.  "
            "Expected one int but got read %zd ints.\n", ints);
    exit(1);
  }
  return ntohl(buf);
}


///
/// Read a cram int from a buffer
///
static int buf_read_int(const char *buf, size_t *offset) {
  const int *int_ptr = (int*)&buf[*offset];
  *offset += sizeof(int);
  return ntohl(*int_ptr);
}


///
/// Read a cram string from a buffer
///
static char *buf_read_string(const char *buf, size_t *offset) {
  size_t size = buf_read_int(buf, offset);
  char *string = strndup(&buf[*offset], size);
  *offset += size;
  return string;
}


static size_t get_cram_buffer_size() {
  const char *bufsize_string = getenv("CRAM_BUFFER_SIZE");
  if (!bufsize_string) {
    return LUSTRE_BUFFER_SIZE;
  }

  int rank;
  PMPI_Comm_rank(MPI_COMM_WORLD, &rank);

  char *endptr;
  size_t bufsize = strtoll(bufsize_string, &endptr, 10);
  if (*bufsize_string && *endptr == '\0') {
    if (rank == 0) {
      fprintf(stderr, "Using CRAM_BUFFER_SIZE=%d.\n", bufsize);
    }
    return bufsize;
  } else {
    if (rank == 0) {
      fprintf(stderr, "Warning: Invalid value for CRAM_BUFFER_SIZE: %s.  "
              "Using default of %d", bufsize_string, LUSTRE_BUFFER_SIZE);
    }
    return LUSTRE_BUFFER_SIZE;
  }
}


// ------------------------------------------------------------------------
// Public cram file interface
// ------------------------------------------------------------------------

bool cram_file_open(const char *filename, cram_file_t *file) {
  file->fd = fopen(filename, "r");
  if (file->fd == NULL) {
    return false;
  }

  // Try to use a large buffer to read the file fast.
  setvbuf(file->fd, NULL, _IOFBF, get_cram_buffer_size());

  // check magic number at start of header
  int magic = file_read_int(file);
  if (magic != MAGIC) {
    fprintf(stderr, "Error: %s is not a cram file!", filename);
    return false;
  }

  // read rest of header after magic check.
  file->version      = file_read_int(file);
  file->num_jobs     = file_read_int(file);
  file->total_procs  = file_read_int(file);
  file->max_job_size = file_read_int(file);

  file->cur_job_record_size = 0;
  file->cur_job_procs = 0;
  file->cur_job_id = -1;

  return true;
}


void cram_file_close(const cram_file_t *file) {
  fclose(file->fd);
}


bool cram_file_has_more_jobs(const cram_file_t *file) {
  return file->cur_job_id < (file->num_jobs - 1);
}


///
/// Helper for cram_job_decompress -- does the real work.
///
static inline void decompress(const cram_job_t *base,
                              int num_missing, const char **missing,
                              int num_changed, const char **changed_keys,
                              const char **changed_vals,
                              cram_job_t *job) {
  // Figure out how many environemnt variables overlap with base
  int num_overlap = 0;
  for (int i=0; i < num_changed; i++) {
    if (index_of(base->num_env_vars, base->keys, changed_keys[i])) {
      num_overlap++;
    }
  }

  // Allocate space for new vars.
  job->num_env_vars = base->num_env_vars
    + num_changed - num_overlap - num_missing;
  job->keys = malloc(job->num_env_vars * sizeof(const char*));
  job->values = malloc(job->num_env_vars * sizeof(const char*));

  // Merge base and changed values into job, removing missing keys.
  // These arrays are assumed to be sorted, so we can march through
  // them in O(n) time.
  size_t bx=0, cx=0, mx=0, jx=0;
  while (jx < job->num_env_vars) {
    int cmp = -1;
    if (cx < num_changed) {
      cmp = strcmp(base->keys[bx], changed_keys[cx]);
    }

    if (cmp == 0) {
      // base & changed are the same, take the changed value and skip base.
      job->keys[jx]   = strdup(changed_keys[cx]);
      job->values[jx] = strdup(changed_vals[cx]);
      bx++; cx++; jx++;

    } else if (cmp < 0) {
      if (mx < num_missing && strcmp(base->keys[bx], missing[mx]) == 0) {
        // This key is missing in job; skip it.
        bx++; mx++;

      } else {
        // base < changed: this key is preserved in the new job.
        job->keys[jx]   = strdup(base->keys[bx]);
        job->values[jx] = strdup(base->values[bx]);
        bx++; jx++;
      }

    } else { /* cmp > 0 */
      // This key was added, and isn't in base. Take it from changed.
      job->keys[jx]   = strdup(changed_keys[cx]);
      job->values[jx] = strdup(changed_vals[cx]);
      cx++; jx++;
    }
  }
}


void cram_job_decompress(const char *job_record,
                         const cram_job_t *base, cram_job_t *job) {
  // start at beginning of job record.
  size_t offset = 0;

  // num_procs
  job->num_procs = buf_read_int(job_record, &offset);

  // working directory
  job->working_dir = buf_read_string(job_record, &offset);

  // command line arguments
  int num_args = buf_read_int(job_record, &offset);
  job->num_args = num_args;
  job->args = (const char**) malloc(num_args * sizeof(char*));

  for (int i=0; i < num_args; i++) {
    job->args[i] = buf_read_string(job_record, &offset);
  }

  // Subtracted environment variables are not in this job but are
  // in the base job.
  int num_subtracted = buf_read_int(job_record, &offset);

  const char **subtracted_env_vars = NULL;
  if (num_subtracted) {
    // If there is no base job, then there can't be any subtracted vars.
    if (!base) {
      fprintf(stderr, "Cannot decompress this job without a base job!\n");
      PMPI_Abort(MPI_COMM_WORLD, 1);
    }

    subtracted_env_vars = malloc(num_subtracted * sizeof(const char*));
    for (int i=0; i < num_subtracted; i++) {
      subtracted_env_vars[i] = buf_read_string(job_record, &offset);
    }
  }

  // Changed environemnt vars were either added to the base or they're different
  // in job from in base.
  int num_changed = buf_read_int(job_record, &offset);

  const char **changed_keys = malloc(num_changed * sizeof(const char*));
  const char **changed_vals = malloc(num_changed * sizeof(const char*));
  for (int i=0; i < num_changed; i++) {
    changed_keys[i] = buf_read_string(job_record, &offset);

    changed_vals[i] = buf_read_string(job_record, &offset);
  }

  if (base) {
    // Apply diffs to the base job and write result into job.
    decompress(base, num_subtracted, subtracted_env_vars,
               num_changed, changed_keys, changed_vals, job);

    // free temporaries
    free_string_array(num_subtracted, subtracted_env_vars);
    free_string_array(num_changed, changed_keys);
    free_string_array(num_changed, changed_vals);

  } else {
    // If there is no base, we just copy the changed vals (first job)
    job->num_env_vars = num_changed;
    job->keys         = changed_keys;
    job->values       = changed_vals;
  }
}


bool cram_file_next_job(cram_file_t *file, char *job_record) {
  int job_record_size = file_read_int(file);
  if (job_record_size > file->max_job_size) {
    fprintf(stderr, "Error: Invalid job record size: %d > %d",
            job_record_size, file->max_job_size);
    return false;
  }

  file->cur_job_record_size = job_record_size;
  size_t bytes = fread(job_record, 1, job_record_size, file->fd);
  if (bytes != job_record_size) {
    fprintf(stderr, "Error: Expected to read %d bytes, but got %zd\n",
            job_record_size, bytes);
    return false;
  }

  size_t offset = 0;
  file->cur_job_procs = buf_read_int(job_record, &offset);
  file->cur_job_id++;

  return true;
}


void cram_file_bcast_jobs(cram_file_t *file, int root, cram_job_t *job, int *id,
                          MPI_Comm comm) {
  int rank, size;
  PMPI_Comm_rank(comm, &rank);
  PMPI_Comm_size(comm, &size);

  // check total procs and grab the max job size.
  int max_job_size;
  if (rank == root) {
    if (file->total_procs > size) {
      fprintf(stderr, "Error: This cram file requires %d processes, "
              "but this communicator has only %d.\n", file->total_procs, size);
      PMPI_Abort(comm, 1);
    }
    max_job_size = file->max_job_size;
  }

  // bcast max job size
  PMPI_Bcast(&max_job_size, 1, MPI_INT, root, comm);
  char *job_record = malloc(max_job_size);

  // read in compressed data for first job record
  if (rank == root) {
    if (!cram_file_next_job(file, job_record)) {
      fprintf(stderr, "Error reading job %d from cram file on rank %d\n",
              0, root);
      PMPI_Abort(comm, 1);
    }
  }

  // Bcast and decompress first job.
  cram_job_t first_job;
  PMPI_Bcast(job_record, max_job_size, MPI_CHAR, root, comm);
  cram_job_decompress(job_record, NULL, &first_job);

  // start by sending to the first rank in the second job.
  int cur_rank = first_job.num_procs;
  bool in_first_job = rank < cur_rank;

  if (rank == root) {
    // Root needs to send to all the other jobs
    // Start by iterating through remaining jobs in the file.
    while (cram_file_has_more_jobs(file)) {
      if (!cram_file_next_job(file, job_record)) {
        fprintf(stderr, "Error reading job %d from cram file on rank %d\n",
                file->cur_job_id + 1, root);
        PMPI_Abort(comm, 1);
      }

      // array of requests for all sends we'll do
      // max concurrent peers max number of ranks we'll send to at once.
      int max_requests = MAX_CONCURRENT_PEERS * 2;
      MPI_Request requests[max_requests];

      // iterate through all ranks in this job, incrementing first_rank as we go.
      int end_rank = cur_rank + file->cur_job_procs;
      while (cur_rank < end_rank) {
        int r = 0;
        while (r < max_requests && cur_rank < end_rank) {
          PMPI_Isend(&file->cur_job_id, 1, MPI_INT, cur_rank,
                     CRAM_TAG, comm, &requests[r++]);
          PMPI_Isend(job_record, file->cur_job_record_size, MPI_CHAR, cur_rank,
                     CRAM_TAG, comm, &requests[r++]);
          cur_rank++;
        }
        PMPI_Waitall(r, requests, MPI_STATUSES_IGNORE);
      }
    }

    // send a job id of -1 to any inactive ranks
    int inactive_rank_id = -1;
    for (; cur_rank < size; cur_rank++) {
      PMPI_Send(&inactive_rank_id, 1, MPI_INT, cur_rank, CRAM_TAG, comm);
    }

  } else if (!in_first_job) {
    // Ranks NOT in the first job need to receive their actual job record.
    // Ranks in the first job already have their job record.
    PMPI_Recv(id, 1, MPI_INT, root, CRAM_TAG, comm, MPI_STATUS_IGNORE);
    if (*id >= 0) {
      PMPI_Recv(job_record, max_job_size, MPI_CHAR, root, CRAM_TAG, comm,
                MPI_STATUS_IGNORE);
    }
    cram_job_decompress(job_record, &first_job, job);
  }

  // If this rank is in the first job, then just copy the first job we
  // got from the bcast.  And set the id to zero.
  if (in_first_job) {
    cram_job_copy(&first_job, job);
    *id = 0;
  }

  // Can free the first job now b/c we don't need it.
  cram_job_free(&first_job);
}


static void arg_copy(const cram_job_t *job,
                     int *dest_argc, const char ***dest_argv) {

    *dest_argc = job->num_args;
    *dest_argv = malloc(job->num_args * sizeof(char**));
    for (int i=0; i < job->num_args; i++) {
        (*dest_argv)[i] = strdup(job->args[i]);
    }
}


void cram_job_setup(const cram_job_t *job, int *argc, const char ***argv) {
  // change working directory
  chdir(job->working_dir);

  // save argv[0] so that we can use it for the first arg.
  const char *exe_name = NULL;
  if (*argc > 0 && *argv) {
    exe_name = (*argv)[0];
  }

  // Replace command line arguments with those of the job.
  arg_copy(job, argc, argv);

  // set argv[0] to the actual exe name
  if (strcmp(job->args[0], CRAM_DEFAULT_EXE) == 0) {
    if (exe_name) {
      free((void*)(*argv)[0]);
      (*argv)[0] = strdup(exe_name);
    }
  }

  // Also copy arguments into globals for Fortran arg interceptors to access.
  arg_copy(job, &cram_argc, &cram_argv);

  // Set environment variables based on the job's key/val pairs.
  for (int i=0; i < job->num_env_vars; i++) {
    setenv(job->keys[i], job->values[i], 1);
  }
}


void cram_job_free(cram_job_t *job) {
  free((char*)job->working_dir);
  free_string_array(job->num_args, job->args);
  free_string_array(job->num_env_vars, job->keys);
  free_string_array(job->num_env_vars, job->values);
}


void cram_job_copy(const cram_job_t *src, cram_job_t *dest) {
  dest->num_procs    = src->num_procs;
  dest->working_dir  = strdup(src->working_dir);

  dest->num_args     = src->num_args;
  dest->args         = dup_string_array(src->num_args, src->args);

  dest->num_env_vars = src->num_env_vars;
  dest->keys         = dup_string_array(src->num_env_vars, src->keys);
  dest->values       = dup_string_array(src->num_env_vars, src->values);
}


void cram_job_print(const cram_job_t *job) {
  printf("  Num procs: %d\n", job->num_procs);
  printf("  Working dir: %s\n", job->working_dir);
  printf("  Arguments:\n");

  printf("      ");
  for (int i=0; i < job->num_args; i++) {
    if (i > 0) printf(" ");
    printf("%s", job->args[i]);
  }
  printf("\n");

  printf("  Environment:\n");
  for (int i=0; i < job->num_env_vars; i++) {
    printf("      '%s' : '%s'\n", job->keys[i], job->values[i]);
  }
}


void cram_file_cat(cram_file_t *file) {
  printf("Number of Jobs:   %12d\n", file->num_jobs);
  printf("Total Procs:      %12d\n", file->total_procs);
  printf("Cram version:     %12d\n", file->version);
  printf("Max job record:   %12d\n", file->max_job_size);
  printf("\n");
  printf("Job information:\n");

  if (!cram_file_has_more_jobs(file)) {
    return;
  }

  // space for raw, compressed job record.
  char *job_record = malloc(file->max_job_size);

  // First job is special because we don't have to decompress
  cram_job_t first_job;
  cram_file_next_job(file, job_record);
  cram_job_decompress(job_record, NULL, &first_job);

  // print first job
  printf("Job %d:\n", file->cur_job_id);
  cram_job_print(&first_job);

  // Rest of jobs are based on first job.
  cram_job_t job;
  while (cram_file_has_more_jobs(file)) {
    cram_file_next_job(file, job_record);
    cram_job_decompress(job_record, &first_job, &job);

    // print each subsequent job
    printf("Job %d:\n", file->cur_job_id);
    cram_job_print(&job);

    cram_job_free(&job);
  }

  free(job_record);
  cram_job_free(&first_job);
}
